lib/internal/util/comparisons.js

```diff
@@ -78,22 +101,37 @@ function strictDeepEqual(val1, val2, memos) {
   if (val1Tag !== val2Tag) {
     return false;
   }
-  if (Object.getPrototypeOf(val1) !== Object.getPrototypeOf(val2)) {
+  if (getPrototypeOf(val1) !== getPrototypeOf(val2)) {
     return false;
   }
   if (val1Tag === '[object Array]') {
     // Check for sparse arrays and general fast path
-    if (val1.length !== val2.length)
+    if (val1.length !== val2.length) {
+      return false;
+    }
+    const keys = objectKeys(val1);
+    if (keys.length !== objectKeys(val2).length) {
       return false;
-    // Skip testing the part below and continue with the keyCheck.
-    return keyCheck(val1, val2, true, memos);
+    }
+    // Fast path for non sparse arrays (no key comparison for indices
+    // properties).
+    if (val1.length === keys.length) {
+      if (keys.length === 0 || keys[val1.length - 1] === `${val1.length - 1}`) {
+        return keyCheck(val1, val2, kStrict, memos, kIsArray, []);
+      }
+    } else if (keys.length > val1.length &&
+               keys[val1.length - 1] === `${val1.length - 1}`) {
+      const minimalKeys = keys.slice(val1.length);
+      return keyCheck(val1, val2, kStrict, memos, kIsArray, minimalKeys);
+    }
+    // Only set this to kIsArray in case the array is not sparse!
+    return keyCheck(val1, val2, kStrict, memos, kNoIterator, keys);
   }
   if (val1Tag === '[object Object]') {
-    // Skip testing the part below and continue with the keyCheck.
-    return keyCheck(val1, val2, true, memos);
+    return keyCheck(val1, val2, kStrict, memos, kNoIterator);
   }
   if (isDate(val1)) {
-    if (val1.getTime() !== val2.getTime()) {
+    if (getTime(val1) !== getTime(val2)) {
```

Comment:
If someone subclasses `Date` and they have an overridden `getTime()` would this work?

If not, I am still +1 for it, but it would require this to be a very explicit semver-major.

Reply:
I'll change it to a separate commit.